astroid/cache.py

```diff
@@ -0,0 +1,104 @@
+""" Caching utilities used in various places."""
+
+import typing
+from collections import OrderedDict
+from weakref import WeakSet
+
+import wrapt
+
+LRU_CACHE_CAPACITY = 128
+
+K = typing.TypeVar("K")
+V = typing.TypeVar("V")
+
+
+class LRUCache(typing.Generic[K, V]):
+    """An LRU cache that keeps track of its instances."""
+
+    instances: "WeakSet[LRUCache[typing.Any, typing.Any]]" = WeakSet()
+
+    def __init__(self) -> None:
+        self.cache: typing.OrderedDict[K, V] = OrderedDict()
+        LRUCache.instances.add(self)
+
+    def __setitem__(self, key: K, value: V) -> None:
+        self.cache[key] = value
+
+        if len(self.cache) > LRU_CACHE_CAPACITY:
+            self.cache.popitem(last=False)
+
+    def __getitem__(self, key: K) -> V:
+        if key in self.cache:
+            self.cache.move_to_end(key)
+
+        return self.cache[key]
+
+    def __contains__(self, key: K) -> bool:
+        return key in self.cache
+
+    def clear(self) -> None:
+        self.cache.clear()
+
+    @classmethod
+    def clear_all(cls) -> None:
+        """Clears all LRUCache instances."""
+        for cache in cls.instances:
+            cache.clear()
+
+
+def lru_cache_astroid(arg=None):
+    """A decorator to cache the results of a function. Similar to
+    functools.lru_cache but uses astroid.cache.LRUCache as its internal cache.
+    """
+    cache = LRUCache()
+
+    @wrapt.decorator
+    def decorator(func, instance, args, kwargs):
```

Comment:
I'm not totally sure, but I think we can a bit better typing-wise based on the documentation [here](https://wrapt.readthedocs.io/en/latest/decorators.html).

`func` is probably a `Callable`? `Instance` `Any`? And `args` and `kwargs` something like `List` and `Dict`. If we can't narrow them down we should at least add `Any` (same goes for the return type) as `mypy` will eventually complain about coming across an untyped function definition.

Reply:
Sure, I've added the typings!

Reply:
You can add a `type: ignore[misc] # wrapt.decorator is untyped` on both lines, we can't do much about it sadly.